lib/pdn/pdn.c

```diff
@@ -0,0 +1,457 @@
+	struct pdn *pdn;
+
+	pdn = pdn_find(cid);
+	if (!pdn) {
+		return -EINVAL;
+	}
+
+	err = snprintf(at_buf, sizeof(at_buf), "AT+CGDCONT=%u", cid);
+	if (err < 0 || err >= sizeof(at_buf)) {
+		return -ENOMEM;
+	}
+
+	err = at_cmd_write(at_buf, NULL, 0, NULL);
+	if (err) {
+		return err;
+	}
+
+	pdn->context_id = CID_UNASSIGNED;
+	pdn->callback = NULL;
+
+	return 0;
+}
+
+static int cgact(uint8_t cid, bool activate)
+{
+	int err;
+
+	err = snprintf(at_buf, sizeof(at_buf), "AT+CGACT=%u,%u",
+		       activate ? 1 : 0, cid);
+	if (err < 0 || err >= sizeof(at_buf)) {
+		return -ENOMEM;
+	}
+
+	err = at_cmd_write(at_buf, NULL, 0, NULL);
+	if (err) {
+		LOG_WRN("Failed to activate PDN for CID %d, err %d", cid, err);
+		return err;
+	}
+
+	return 0;
+}
+
+int pdn_activate(uint8_t cid, int *esm)
+{
+	int err;
+	int timeout;
+
+	err = cgact(cid, true);
+	if (esm) {
+		timeout = k_sem_take(&notif_sem, K_MSEC(CONFIG_PDN_ESM_TIMEOUT));
+		if (!timeout) {
+			*esm = esm_from_notif;
+		} else if (timeout == -EAGAIN) {
+			*esm = 0;
+		}
+		esm_from_notif = 0;
+	}
+
+	return err;
+}
+
+int pdn_deactivate(uint8_t cid)
+{
+	int err;
+
+	err = cgact(cid, false);
+	if (err) {
+		return err;
+	}
+
+	return 0;
+}
+
+int pdn_id_get(uint8_t cid)
+{
+	int err;
+	char *p;
+	char resp[32];
+
+	err = snprintf(at_buf, sizeof(at_buf), "AT%%XGETPDNID=%u", cid);
+	if (err < 0 || err >= sizeof(at_buf)) {
+		return -ENOMEM;
+	}
+
+	err = at_cmd_write(at_buf, resp, sizeof(resp), NULL);
+	if (err) {
+		LOG_ERR("Failed to read PDN ID for CID %d, err %d", cid, err);
+		return err;
+	}
+
+	p = strchr(resp, ':');
+	if (!p) {
+		return -EBADMSG;
+	}
+
+	return atoi(p + 1);
+}
+
+int pdn_default_apn_get(char *buf, size_t len)
+{
+	int err;
+	char *p, *end;
+
+	err = at_cmd_write("AT+CGDCONT?", at_buf, sizeof(at_buf), NULL);
+	if (err) {
+		LOG_ERR("Failed to read PDP contexts, err %d", err);
+		return err;
+	}
+
+	/* CGDCONT: #,"Family","APN" */
+	p = strstr(at_buf, "CGDCONT: 0");
+	if (!p) {
+		return -EBADMSG;
+	}
+
+	/* Skip to the third double quote `"` */
+	for (size_t i = 0; i < 3; i++) {
+		p = strchr(p, '\"');
+		if (!p) {
+			return -EBADMSG;
+		}
+		p++;
+	}
+
+	/* Find closing quote and replace with \0 */
+	end = strchr(p, '\"');
+	if (!end) {
+		return -EBADMSG;
+	}
+	*end = '\0';
+
+	if (end - p + sizeof('\0') > len) {
+		return -ENOMEM;
+	}
+
+	memcpy(buf, p, end - p + sizeof('\0'));
+
+	return 0;
+}
+
+#if defined(CONFIG_PDN_SYS_INIT)
+static int pdn_sys_init(const struct device *unused)
+{
+	int err;
+
+#if defined(CONFIG_PDN_DEFAULTS_OVERRIDE)
+	err = pdn_cid_configure(0, CONFIG_PDN_DEFAULT_APN,
+				CONFIG_PDN_DEFAULT_FAM, NULL);
+	if (err) {
+		LOG_ERR("Failed to configure default CID, err %d", err);
+		return -1;
+	}
+
+	err = pdn_cid_auth_set(0, CONFIG_PDN_DEFAULT_AUTH,
+			       CONFIG_PDN_DEFAULT_USERNAME,
+			       CONFIG_PDN_DEFAULT_PASSWORD);
+	if (err) {
+		LOG_ERR("Failed to set auth params for default CID, err %d",
+			err);
+		return -1;
+	}
+#endif
```

Comment:
A comment on what `if` ends here and the `endif` a few lines below would be good

Reply:
Done